yesod-bin/Devel.hs

```diff
@@ -222,6 +227,42 @@ checkDevelFile =
             then return x
             else loop xs
 
+stackSuccessString :: ByteString
+stackSuccessString = "ExitSuccess"
+
+stackFailureString :: ByteString
+stackFailureString = "ExitFailure"
+
+-- We need updateAppPort logic to prevent a race condition.
+-- See https://github.com/yesodweb/yesod/issues/1380
+updateAppPort :: ByteString -> TVar Bool -- ^ Bool to indicate if the
+                                         -- output from stack has
+                                         -- started. False indicate
+                                         -- that it hasn't started
+                                         -- yet.
+              -> TVar Int -> IO ()
+updateAppPort bs buildStarted appPortVar = do
+  hasStarted <- readTVarIO buildStarted
```

Comment:
This looks like a race condition to me, at least in principle. Why not turn the entire function into a single `STM` block and perform the read and modifications atomically?